Automatic rom/obb extractor and rom repacker

fenwaypowers opened the pull request:
I have added three new functions to the script: 

- `automatic_extract_repack()` starts off the process, but mainly is used for making sure the user has ndstool on their system. If they don't, the program will download it for them. 

- `extract()` extracts the mpt files from the US rom and obb file and places them in `en` with ndstool. This function will only be called if the user does not already have all the mpt files required in the `en` folder or if the ja rom has not been extracted.

- `repack()` takes a copy of the extraced JA rom and modifies it by placing the patched mpt files in and then uses ndstool to repack the rom.

This is all to make the process of installing this patch much easier, and to help DQ fans and non DQ fans be able to experience this game with its full features on DS.

These functions work and have been tested on Linux, Windows, and MacOS.

In the event that ndstool does not work on their system, or the functions don't work, the user can skip these functionalities with the argument `--manual`, which lets the script work exactly as it did before.

fenwaypowers committed:
initial commit

not working yet, just testing.

fenwaypowers committed:
obb extractor

fenwaypowers committed:
process is complete

fenwaypowers committed:
small fix

fenwaypowers committed:
repacking finished

fenwaypowers committed:
add new funcs to dqiv_patch.py

add processes from installer.py to dqiv_patch.py.

implement the --manual arg that disables automatic extract and repack features.

fenwaypowers committed:
Delete installer.py

fenwaypowers committed:
fix grammar

fenwaypowers committed:
added ability to download ndstool from within the script

fenwaypowers committed:
ndstool downloader functionality working for linux, presumed working for windows.

fenwaypowers committed:
windows support

fenwaypowers committed:
Update dqiv_patch.py

fenwaypowers committed:
Update dqiv_patch.py

fenwaypowers committed:
Update dqiv_patch.py

if mpt files are already extracted, the program won't extract them again needlessly.

fenwaypowers committed:
Update dqiv_patch.py and README

fenwaypowers committed:
Update README.md

fenwaypowers committed:
Update dqiv_patch.py

fenwaypowers commented:
Thank you! I'm ready for feedback and would like to work through any suggestions you have.

I just realized after I made the pull request that the `extract()` function should probably be updated to allow for only extracting the ja rom in the case of the US rom english files already having been extracted.

concreted commented on dqiv_patch.py:
```diff
@@ -500,5 +514,202 @@ def patch_file_en(filename):
+
+            download = requests.get(ndstool_links[dl_list[selection - 1]])
+
+            if os.path.exists("ndstool") == False:
```
(optional) General comment, usually you can just do `if not X` to check for false conditions - the python style guide recommends not using `== True` or `== False` for boolean comparisons (https://peps.python.org/pep-0008/#programming-recommendations).

fenwaypowers replied:
Makes sense. I think it's just a habit of mine to do `== False`, so I can change this.

concreted commented on dqiv_patch.py:
```diff
@@ -500,5 +514,202 @@ def patch_file_en(filename):
+
+            counter = 1
+            dl_list = []
+            for link in ndstool_links:
```
(optional) Can use `for i, link in enumerate(ndstool_links)` which will set `i` to the index of the item. `i` would be 0-indexed so would have to do something like `print("[" + str(i+1) + "] " + link)`.

concreted commented on dqiv_patch.py:
```diff
@@ -500,5 +514,202 @@ def patch_file_en(filename):
+            must_extract = True
+
+    if os.path.exists("roms/ja") == False:
+        must_extract = True
```
As you pointed out, it seems like the logic to extract ROMs and OBB content is tied together. It would be nice if it could only extract the ROMs or Android content if needed when calling `extract`. Or could have separate functions to extract the ROMs and OBB.

concreted commented on dqiv_patch.py:
```diff
@@ -500,5 +514,202 @@ def patch_file_en(filename):
+
+        print("Extracting " + region + " rom...")
+        subprocess.run(path_to_ndstool + " -x " + roms[region] + " -9 " + path_to_region_folder + "/arm9.bin -7 " + path_to_region_folder + "/arm7.bin -y9 " + path_to_region_folder + "/y9.bin -y7 " + path_to_region_folder + "/y7.bin -t " + path_to_region_folder + "/banner.bin -h " + path_to_region_folder + "/header.bin -d " + path_to_region_folder + "/data -y " + path_to_region_folder + "/overlay ...
+        print("Extraction of " + region + " rom complete.")
```
(optional) I generally prefer using syntax like `f"Extract of {region} rom complete` for string formatting instead of `+`ing strings together. Not a big deal though, I think this is fine :)

concreted commented on dqiv_patch.py:
```diff
@@ -500,5 +514,202 @@ def patch_file_en(filename):
+            try:
+                os.rename("en/" + mpt, "en/" + mpt.split("assets/msg/en/")[1])
+            except:
+                pass
```
Not sure if it is a problem if this rename doesn't happen? If so would rather raise the exception so user is aware of it.

fenwaypowers replied:
right. this was to fix an issue with windows. on linux when you rename a file and the target path for that file already exists, the file that was already there is overwritten, but on windows it raises an error.

I think this `try` statement can be completely removed due to the fact that extracting the rom when the mpt files are already in `en` is no longer something that happens.

concreted commented on dqiv_patch.py:
```diff
@@ -500,5 +514,202 @@ def patch_file_en(filename):
+    #remove the repack folder
+    shutil.rmtree(path_to_repack)
+
+    #remove the ndstool zip
```
(optional) General comment, for Python comments I prefer to add a space after the `#` e.g. `# remove the ndstool zip`, this is recommended by the styleguide (https://peps.python.org/pep-0008/#block-comments) and I think it's more readable.

concreted commented on dqiv_patch.py:
```diff
@@ -500,5 +514,202 @@ def patch_file_en(filename):
+
 if __name__ == "__main__":
-    main()
+    main()
```
Let's leave the newline at the end of the file (https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline) - I don't think it's usually big deal in practice but I don't like seeing the Github warning :) 
![image](https://user-images.githubusercontent.com/4016897/223312412-bfebf3a9-e3a9-4d6d-85cf-821edd76558b.png)

concreted commented on .gitignore:
```diff
@@ -1,2 +1,9 @@
+*.nds
+*.obb
+test.py
+ndstool
```
Same comment re: newline at end of file as `dqiv_patch.py`

concreted reviewed (commented):
Overall looks good and I super appreciate you adding this functionality to the script. Left some comments - the ones starting with `(optional)` are related to style, I am not too hung up on style for this script so treat those as optional.

concreted commented on dqiv_patch.py:
```diff
@@ -500,5 +514,202 @@ def patch_file_en(filename):
+        ndstool_found = True
+    else:
+        for possible_path in ["ndstool/ndstool", "ndstool/ndstool.exe"]:
+            if os.path.exists(possible_path):
+                ndstool_found = True
+                if possible_path.endswith(".exe"):
+                    path_to_ndstool = "ndstool\\ndstool.exe"
```
(optional) I've found pathlib (https://docs.python.org/3/library/pathlib.html) to be pretty useful and standardized for filesystem related stuff like making dirs, finding if files exist, and checking extensions. No need to use it in this script but just thought I'd mention :)

fenwaypowers committed:
A few fixes

fenwaypowers committed:
Update dqiv_patch.py

Program only extracts when it needs to. JA rom is set to always automatically extract, but if it's already extracted, ndstool will recognize this itself and won't extract the rom.

fenwaypowers committed:
patched files now go to patched folder

fenwaypowers committed:
Update README.md

fenwaypowers commented:
Hey! So I've been busy lately but have got to work on making it so the program extracts only what is needed, as well as some previous formatting issues.

This has lead to the splitting of `extract()` into three functions:
- `extract_roms()` : responsible for extracting each NDS rom as needed.
- `move_nds_mpt()` : responsible for moving the US NDS English mpt files from the extracted folder to the `en` folder as needed.
- `extract_obb()` : responsible for extracting the obb mpt files to `en`.

Please take a look at my most recent commit, and let me know if there's anything else that needs addressing.

concreted commented:
Looks good, merged! Thanks again for contributing this feature and improving the user-friendliness of this script.
